File: Easy/JAVASCRIPT/PalindromeCheck.js
// O(n^2) time | O(n) space
function isPalindrome1(str) {
    if (str === null)
        return false;

    let input = "";

    for (let i = str.length - 1; i >= 0; i--)
        // input += str[i];
        input += str.charAt(i);

    return str === input;
}

// O(n) time | O(n) space
function isPalindrome2(str) {
    let reversed = str.split("").reverse().join("");
    return reversed === str ;
}

// O(n) time | O(n) space
function isPalindrome3(str) {
    let input = [];

    for (let i = str.length - 1; i >= 0; i--)
    input.push(str.charAt(i));

    return str === input.join('');
}

    // O(n) time | O(n) space
function isPalindrome4(str) {
    if (str == null)
        return false;

    return isPalindromeHelper(str, 0);
}

function isPalindromeHelper(str,i) {
    let j = str.length  - (1 + i);
    return i >= j ? true : str.charAt(i) == str.charAt(j) && isPalindromeHelper(str, i + 1);
}

// O(n) time | O(1) space
function isPalindrome5(str) {
    if (str == null)
        return false;

    var left = 0;
    var right = str.length - 1;

    while (left < right) {
        if (str.charAt(left++) != str.charAt(right--))
            return false;
    }
    return true;
}

const str = "1234321";
const result = isPalindrome3(str);
console.log(result);

File: Easy/PYTHON/PalindromeCheck.py
def is_palindrome_1(str):
    if str is None:
        return False

    input = ""
    for i in range(len(str)-1, -1, -1):
        input += str[i]

    return str == input


#  O(n) time | O(n) space
def is_palindrome_2(str):
    reversed = str[::-1]
    return reversed == str

# O(n) time | O(n) space


def is_palindrome_3(str):
    input = []

    for i in range(len(str)-1, -1, -1):
        input.append(str[i])

    return str == ''.join(input)

# O(n) time | O(n) space


def is_palindrome_4(str):
    if (str is None):
        return False

    return is_palindrome_4_helper(str, 0)


def is_palindrome_4_helper(str, i):
    j = len(str) - (1 + i)
    return True if (i >= j) else str[i] == str[j] and is_palindrome_4_helper(str, i + 1)


# O(n) time | O(1) space
def is_palindrome_5(str):
    if (str is None):
        return False

    left = 0
    right = len(str) - 1

    while (left < right):
        if str[left] != str[right]:
            return False
        left += 1
        right -= 1

    return True


str = "1234321"
result = is_palindrome_4(str)
print(result)
